### ISSUE TITLE
Failed to update lastScheduleTime

### ISSUE THREAD

**Issue by randrusiak** [chaos-mesh/chaos-mesh#3884]
## Bug Report

**What version of Kubernetes are you using?**

v1.23.10

**What version of Chaos Mesh are you using?**

Controller manager Version: version.Info{GitVersion:"chart-2.5.0", GitCommit:"10bc9220553c6f1431e1ecb111113bc84f3cf2ab", BuildDate:"2022-11-22T11:35:49Z", GoVersion:"go1.18", Compiler:"gc", Platform:"linux/amd64"}

**What did you do? / Minimal Reproducible Example**

Create example schedule manifest:

```
kind: Schedule
apiVersion: chaos-mesh.org/v1alpha1
metadata:
  namespace: test
  name: test-nginx
spec:
  schedule: '* * * * *'
  startingDeadlineSeconds: null
  concurrencyPolicy: Forbid
  type: PodChaos
  podChaos:
    selector:
      namespaces:
        - test
      labelSelectors:
        app: test-nginx
    mode: one
    action: pod-kill
```

**What did you expect to see?**

I assume that one random pod will be killed every minute.

**What did you see instead?**

Sometimes more than one pod is killed because schedule-cron cannot update lastScheduleTime (.status.time). 
It causes the creation of multiple experiments because the chaos-controller assume that no experiment has run before. 

Error from events:

```
Failed to update lastScheduleTime: Operation cannot be fulfilled on schedules.chaos-mesh.org "test-nginx": the object has been modified; please apply your changes to the latest version and try again
```

Error from controller-manager output:
```
2023-01-05T14:55:00.754Z	ERROR	schedule-cron	cron/controller.go:181	fail to update	{"error": "Operation cannot be fulfilled on schedules.chaos-mesh.org \"test-nginx\": the object has been modified; please apply your changes to the latest version and try again"}
github.com/chaos-mesh/chaos-mesh/controllers/schedule/cron.(*Reconciler).Reconcile
	/home/runner/work/chaos-mesh/chaos-mesh/controllers/schedule/cron/controller.go:181
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	/tmp/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:114
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/tmp/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:311
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/tmp/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/tmp/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:227
```

**Comment by STRRL** [comment]
Hi @randrusiak, it seems the resource `Schedule` is failed to updated because of the conflicts.

Could you provide us the output of `kubectl get podchaos -o wide` when you reproduce this issue?

**Comment by STRRL** [comment]
the number of podchaos should be same as the pasted time in minus you created the `Schedule`.

**Comment by randrusiak** [comment]
When the issue is reproduced I noticed that multiple PodChaos are created.
And what is interesting they are changed multiple times withing **~1s:**

`k get podchaos -A --watch`

```
test   test-nginx-slsjh   0s
test   test-nginx-slsjh   0s
test   test-nginx-slsjh   0s
test   test-nginx-slsjh   0s
test   test-nginx-slsjh   0s
test   test-nginx-slsjh   0s
test   test-nginx-slsjh   1s
test   test-nginx-slsjh   1s
test   test-nginx-g9j9s   0s
test   test-nginx-slsjh   1s
test   test-nginx-slsjh   1s
test   test-nginx-g9j9s   0s
test   test-nginx-g9j9s   0s
test   test-nginx-g9j9s   0s
test   test-nginx-g9j9s   1s
test   test-nginx-g9j9s   1s
test   test-nginx-g9j9s   1s
test   test-nginx-6876c   0s
test   test-nginx-g9j9s   1s
test   test-nginx-g9j9s   1s
test   test-nginx-g9j9s   2s
test   test-nginx-6876c   1s
test   test-nginx-6876c   1s
test   test-nginx-6876c   1s
test   test-nginx-6876c   2s
test   test-nginx-6876c   2s
test   test-nginx-gnmw2   0s
test   test-nginx-6876c   2s
test   test-nginx-gnmw2   0s
test   test-nginx-gnmw2   0s
test   test-nginx-gnmw2   0s
test   test-nginx-gnmw2   1s
test   test-nginx-6876c   3s
test   test-nginx-gnmw2   1s
test   test-nginx-6876c   3s
test   test-nginx-6876c   3s
test   test-nginx-gnmw2   1s
test   test-nginx-gnmw2   1s
test   test-nginx-gnmw2   2s
test   test-nginx-gnmw2   2s
```
`kubectl get events`  
```
13s         Normal    Spawned             schedule/test-nginx                                           Create new object: test-nginx-slsjh
12s         Warning   Failed              schedule/test-nginx                                           Failed to update lastScheduleTime: Operation cannot be fulfilled on schedules.chaos-mesh.org "test-nginx": the object has been modified; please apply your changes to the latest version and try again
12s         Normal    Spawned             schedule/test-nginx                                           Create new object: test-nginx-g9j9s
11s         Warning   Failed              schedule/test-nginx                                           Failed to update lastScheduleTime: Operation cannot be fulfilled on schedules.chaos-mesh.org "test-nginx": the object has been modified; please apply your changes to the latest version and try again
11s         Normal    Spawned             schedule/test-nginx                                           Create new object: test-nginx-6876c
10s         Warning   Failed              schedule/test-nginx                                           Failed to update lastScheduleTime: Operation cannot be fulfilled on schedules.chaos-mesh.org "test-nginx": the object has been modified; please apply your changes to the latest version and try again
10s         Warning   Forbid              schedule/test-nginx                                           Forbid spawning new job because: test-nginx-6876c is still running
9s          Warning   Forbid              schedule/test-nginx                                           Forbid spawning new job because: test-nginx-6876c is still running
9s          Warning   Forbid              schedule/test-nginx                                           Forbid spawning new job because: test-nginx-6876c is still running
9s          Normal    Spawned             schedule/test-nginx                                           Create new object: test-nginx-gnmw2
9s          Normal    Updated             schedule/test-nginx                                           Successfully update lastScheduleTime of resource
```

**Comment by randrusiak** [comment]
@STRRL any ideas?

**Comment by AleksanderWWW** [comment]
Is there any update here? It's blocking the usage of the project in my case - when encountering this error the schedule goes on a pod-killing frenzy of retries frying my cluster, when all I wanted is to kill a random pod every hour or so.

@STRRL

**Comment by STRRL** [comment]
sorry for the late response, I would take a look about the racing states

**Comment by STRRL** [comment]
>Is there another process which updates the schedule object, or maybe the object data is (wrongly) cached?
Why do we copy the schedule object and update it? Maybe we can only perform patch request which updates only status.time field.
Or alternatively, we should clear some metadata before performing update request

We follow the kubernetes pattern to update objects, so there did some other process updated it, so it failed  to update the data.

But it seems that would not call another reconciliation of creating podchaos, let me take a look.

**Comment by anthosz** [comment]
Hi,

I can reproduce the same issue randomly.

I have 3 sentinel redis, trying to kill randomly one by one but due to the paralell kill, sometimes it kill the quorum :/

**Comment by STRRL** [comment]
> trying to kill randomly one by one but due to the paralell kill, sometimes it kill the quorum

I think it might be off-topic to this thread. Would you like to open a new issue with your definition of PodChaos? :heart:

**Comment by anthosz** [comment]
> > trying to kill randomly one by one but due to the paralell kill, sometimes it kill the quorum
> 
> I think it might be off-topic to this thread. Would you like to open a new issue with your definition of PodChaos? ❤️

I think it's related to this topic, it was exactly the same issue: I see this kind of events each time that there is parallel pods killed in parallel:
Failed to update lastScheduleTime: Operation cannot be fulfilled on schedules.chaos-mesh.org "xxxxx": the object has been modified; please apply your changes to the latest version and try again.

At the end, randomly, instead to terminate pod one by one, when this error spawn, it seems it run 2 times (even if it "works" because the first request seems to do the job) and it kill 2 pods instead 1.

Here is the related definition:
```
kind: Schedule
apiVersion: chaos-mesh.org/v1alpha1
metadata:
  namespace: chaos-mesh
  name: redis-killer
  annotations:
    experiment.chaos-mesh.org/pause: 'false'
spec:
  schedule: '@every 45s'
  startingDeadlineSeconds: null
  concurrencyPolicy: Forbid
  historyLimit: 20
  type: PodChaos
  podChaos:
    selector:
      namespaces:
        - default
      labelSelectors:
        app.kubernetes.io/component: node
      podPhaseSelectors:
        - Running
    mode: fixed
    value: '1'
    action: pod-kill
    duration: 45s
```
